### PR TITLE
Fix excessive IsEnoughWorkspace warning messages after #2947

### LINKED ISSUE #2981: Excessive warning messages on workspace provided and required (IsEnoughWorkspace) after #2947
https://github.com/ROCm/MIOpen/blob/92a3eb11d89ca86ea471f2d74dfbc2044aa04d41/src/conv/solver_finders.cpp#L371-L373

here is the observation after #2947
```
:/var/lib/jenkins/pytorch-micro-benchmarking# python3 micro_benchmarking_pytorch.py --distributed_dataparallel --dist-backend nccl --network resnet152 --batch-size 128 --iterations 100 --device_ids=0 --rank 0 --world-size 1 --dist-url tcp://127.0.0.1:4332
Initializing process group...
Rendezvous complete. Created process group...
INFO: Rank 0 running distributed_dataparallel on devices: [0]
INFO: running forward and backward for warmup.
MIOpen(HIP): Warning [IsEnoughWorkspace] [GetSolutions] Solver <ConvAsmImplicitGemmGTCDynamicFwdXdlopsNHWC>, workspace required: 488149760, provided ptr: 0 size: 0
MIOpen(HIP): Warning [IsEnoughWorkspace] [GetSolutions] Solver <ConvAsmImplicitGemmGTCDynamicFwdXdlopsNHWC>, workspace required: 257490944, provided ptr: 0 size: 0
MIOpen(HIP): Warning [IsEnoughWorkspace] [GetSolutions] Solver <GemmFwd1x1_0_2>, workspace required: 308281344, provided ptr: 0 size: 0
MIOpen(HIP): Warning [IsEnoughWorkspace] [GetSolutions] Solver <ConvAsmImplicitGemmGTCDynamicFwdXdlopsNHWC>, workspace required: 308281344, provided ptr: 0 size: 0
MIOpen(HIP): Warning [IsEnoughWorkspace] [GetSolutions] Solver <ConvAsmImplicitGemmGTCDynamicFwdXdlopsNHWC>, workspace required: 130809856, provided ptr: 0 size: 0
MIOpen(HIP): Warning [IsEnoughWorkspace] [GetSolutions] Solver <ConvAsmImplicitGemmGTCDynamicFwdXdlopsNHWC>, workspace required: 128450560, provided ptr: 0 size: 0
MIOpen(HIP): Warning [IsEnoughWorkspace] [GetSolutions] Solver <ConvAsmImplicitGemmGTCDynamicFwdXdlopsNHWC>, workspace required: 308281344, provided ptr: 0 size: 0
MIOpen(HIP): Warning [IsEnoughWorkspace] [GetSolutions] Solver <GemmFwd1x1_0_2>, workspace required: 154140672, provided ptr: 0 size: 0
MIOpen(HIP): Warning [IsEnoughWorkspace] [GetSolutions] Solver <ConvAsmImplicitGemmGTCDynamicFwdXdlopsNHWC>, workspace required: 128450560, provided ptr: 0 size: 0
MIOpen(HIP): Warning [IsEnoughWorkspace] [GetSolutions] Solver <ConvAsmImplicitGemmGTCDynamicFwdXdlopsNHWC>, workspace required: 53739520, provided ptr: 0 size: 0
MIOpen(HIP): Warning [IsEnoughWorkspace] [GetSolutions] Solver <ConvAsmImplicitGemmGTCDynamicFwdXdlopsNHWC>, workspace required: 154140672, provided ptr: 0 size: 0
MIOpen(HIP): Warning [IsEnoughWorkspace] [GetSolutions] Solver <ConvAsmImplicitGemmGTCDynamicFwdXdlopsNHWC>, workspace required: 64225280, provided ptr: 0 size: 0
MIOpen(HIP): Warning [IsEnoughWorkspace] [GetSolutions] Solver <ConvAsmImplicitGemmGTCDynamicFwdXdlopsNHWC>, workspace required: 154140672, provided ptr: 0x7f683ea00000 size: 77070336
MIOpen(HIP): Warning [IsEnoughWorkspace] [GetSolutions] Solver <ConvAsmImplicitGemmGTCDynamicFwdXdlopsNHWC>, workspace required: 64225280, provided ptr: 0 size: 0
MIOpen(HIP): Warning [IsEnoughWorkspace] [GetSolutions] Solver <ConvAsmImplicitGemmGTCDynamicBwdXdlopsNHWC>, workspace required: 64225280, provided ptr: 0 size: 0
MIOpen(HIP): Warning [IsEnoughWorkspace] [GetSolutions] Solver <ConvAsmImplicitGemmGTCDynamicBwdXdlopsNHWC>, workspace required: 64225280, provided ptr: 0 size: 0
MIOpen(HIP): Warning [IsEnoughWorkspace] [GetSolutions] Solver <ConvWinograd3x3MultipassWrW<3-2>>, workspace required: 235929600, provided ptr: 0x7f6845400000 size: 73662464
MIOpen(HIP): Warning [IsEnoughWorkspace] [GetSolutions] Solver <ConvAsmImplicitGemmGTCDynamicBwdXdlopsNHWC>, workspace required: 154140672, provided ptr: 0 size: 0
MIOpen(HIP): Warning [IsEnoughWorkspace] [GetSolutions] Solver <ConvAsmImplicitGemmGTCDynamicWrwXdlopsNHWC>, workspace required: 154140672, provided ptr: 0 size: 0
MIOpen(HIP): Warning [IsEnoughWorkspace] [GetSolutions] Solver <ConvAsmImplicitGemmGTCDynamicBwdXdlopsNHWC>, workspace required: 128450560, provided ptr: 0 size: 0
MIOpen(HIP): Warning [IsEnoughWorkspace] [GetSolutions] Solver <ConvAsmImplicitGemmGTCDynamicWrwXdlopsNHWC>, workspace required: 128450560, provided ptr: 0 size: 0
MIOpen(HIP): Warning [IsEnoughWorkspace] [GetSolutions] Solver <ConvAsmImplicitGemmGTCDynamicBwdXdlopsNHWC>, workspace required: 53739520, provided ptr: 0 size: 0
MIOpen(HIP): Warning [IsEnoughWorkspace] [GetSolutions] Solver <ConvWinograd3x3MultipassWrW<3-6>>, workspace required: 167772160, provided ptr: 0x7f6849e00000 size: 53739520
MIOpen(HIP): Warning [IsEnoughWorkspace] [GetSolutions] Solver <ConvAsmImplicitGemmGTCDynamicBwdXdlopsNHWC>, workspace required: 128450560, provided ptr: 0 size: 0
MIOpen(HIP): Warning [IsEnoughWorkspace] [GetSolutions] Solver <ConvAsmImplicitGemmGTCDynamicWrwXdlopsNHWC>, workspace required: 128450560, provided ptr: 0 size: 0
MIOpen(HIP): Warning [IsEnoughWorkspace] [GetSolutions] Solver <ConvAsmImplicitGemmGTCDynamicWrwXdlopsNHWC>, workspace required: 308281344, provided ptr: 0 size: 0
MIOpen(HIP): Warning [IsEnoughWorkspace] [GetSolutions] Solver <ConvWinograd3x3MultipassWrW<3-2>>, workspace required: 327680000, provided ptr: 0 size: 0
MIOpen(HIP): Warning [IsEnoughWorkspace] [GetSolutions] Solver <ConvAsmImplicitGemmGTCDynamicWrwXdlopsNHWC>, workspace required: 130809856, provided ptr: 0 size: 0
MIOpen(HIP): Warning [IsEnoughWorkspace] [GetSolutions] Solver <ConvAsmImplicitGemmGTCDynamicBwdXdlopsNHWC>, workspace required: 308281344, provided ptr: 0 size: 0
MIOpen(HIP): Warning [IsEnoughWorkspace] [GetSolutions] Solver <ConvAsmImplicitGemmGTCDynamicWrwXdlopsNHWC>, workspace required: 308281344, provided ptr: 0 size: 0
MIOpen(HIP): Warning [IsEnoughWorkspace] [GetSolutions] Solver <ConvAsmImplicitGemmGTCDynamicWrwXdlopsNHWC>, workspace required: 256901120, provided ptr: 0 size: 0
MIOpen(HIP): Warning [IsEnoughWorkspace] [GetSolutions] Solver <ConvAsmImplicitGemmGTCDynamicBwdXdlopsNHWC>, workspace required: 103350272, provided ptr: 0 size: 0
MIOpen(HIP): Warning [IsEnoughWorkspace] [GetSolutions] Solver <ConvWinograd3x3MultipassWrW<3-6>>, workspace required: 213909504, provided ptr: 0x7f6b33e00000 size: 103350272
MIOpen(HIP): Warning [IsEnoughWorkspace] [GetSolutions] Solver <ConvAsmImplicitGemmGTCDynamicWrwXdlopsNHWC>, workspace required: 256901120, provided ptr: 0 size: 0
MIOpen(HIP): Warning [IsEnoughWorkspace] [GetSolutions] Solver <ConvAsmImplicitGemmGTCDynamicBwdXdlopsNHWC>, workspace required: 616562688, provided ptr: 0 size: 0
MIOpen(HIP): Warning [IsEnoughWorkspace] [GetSolutions] Solver <ConvAsmBwdWrW1x1>, workspace required: 102760448, provided ptr: 0 size: 0
MIOpen(HIP): Warning [IsEnoughWorkspace] [GetSolutions] Solver <ConvAsmImplicitGemmGTCDynamicWrwXdlopsNHWC>, workspace required: 616562688, provided ptr: 0 size: 0
MIOpen(HIP): Warning [IsEnoughWorkspace] [GetSolutions] Solver <ConvWinograd3x3MultipassWrW<3-2>>, workspace required: 643891200, provided ptr: 0 size: 0
MIOpen(HIP): Warning [IsEnoughWorkspace] [GetSolutions] Solver <ConvAsmImplicitGemmGTCDynamicWrwXdlopsNHWC>, workspace required: 257490944, provided ptr: 0 size: 0
MIOpen(HIP): Warning [IsEnoughWorkspace] [GetSolutions] Solver <ConvAsmImplicitGemmGTCDynamicWrwXdlopsNHWC>, workspace required: 616562688, provided ptr: 0 size: 0
MIOpen(HIP): Warning [IsEnoughWorkspace] [GetSolutions] Solver <ConvAsmImplicitGemmGTCDynamicWrwXdlopsNHWC>, workspace required: 513802240, provided ptr: 0 size: 0
MIOpen(HIP): Warning [IsEnoughWorkspace] [GetSolutions] Solver <ConvAsmImplicitGemmGTCDynamicBwdXdlopsNHWC>, workspace required: 205668352, provided ptr: 0 size: 0
MIOpen(HIP): Warning [IsEnoughWorkspace] [GetSolutions] Solver <ConvWinograd3x3MultipassWrW<3-6>>, workspace required: 420478976, provided ptr: 0 size: 0
MIOpen(HIP): Warning [IsEnoughWorkspace] [GetSolutions] Solver <ConvAsmImplicitGemmGTCDynamicWrwXdlopsNHWC>, workspace required: 205668352, provided ptr: 0 size: 0
MIOpen(HIP): Warning [IsEnoughWorkspace] [GetSolutions] Solver <ConvAsmImplicitGemmGTCDynamicWrwXdlopsNHWC>, workspace required: 513802240, provided ptr: 0 size: 0
MIOpen(HIP): Warning [IsEnoughWorkspace] [GetSolutions] Solver <ConvAsmImplicitGemmGTCDynamicWrwXdlopsNHWC>, workspace required: 205520896, provided ptr: 0 size: 0
MIOpen(HIP): Warning [IsEnoughWorkspace] [GetSolutions] Solver <ConvOclBwdWrW2<16>>, workspace required: 301056, provided ptr: 0 size: 0
MIOpen(HIP): Warning [IsEnoughWorkspace] [GetSolutions] Solver <ConvOclBwdWrW2<8>>, workspace required: 602112, provided ptr: 0 size: 0
MIOpen(HIP): Warning [IsEnoughWorkspace] [GetSolutions] Solver <ConvAsmImplicitGemmGTCDynamicWrwXdlopsNHWC>, workspace required: 488149760, provided ptr: 0 size: 0
INFO: running the benchmark..
OK: finished running benchmark..
--------------------SUMMARY--------------------------
Microbenchmark for network : resnet152
--------This process: rank 0--------
Num devices: 1
Dtype: FP32
Mini batch size [img] : 128
Time per mini-batch : 0.4891260766983032
Throughput [img/sec] : 261.6912205213532

--------Overall (all ranks) (assuming same num/type devices for each rank)--------
Num devices: 1
Dtype: FP32
Mini batch size [img] : 128
Time per mini-batch : 0.4891260766983032
Throughput [img/sec] : 261.6912205213532
```

FYI: @atamazov

### PR DESCRIPTION
**Pull request by atamazov**
- Resolves #2981.

----
[Attribution] @junliume @JehandadKhan
- https://github.com/ROCm/MIOpen/labels/quality
- https://github.com/ROCm/MIOpen/labels/urgency_high
- Proposed reviewers:
  - @JehandadKhan 
  - @junliume
  - @DrizztDoUrden

### CODE REVIEW

File: src/ocl/convolutionocl.cpp [hunk 1]
```diff
@@ -773,20 +779,41 @@ std::vector<miopenConvSolution_t> GetSolutions(const ExecutionContext& ctx,
...
     }
 
+    /// Non-zero InvokeParams means that this function is used in Find to optimize host-side
+    /// performance (see Hybrid Find modes). Note that maxSolutionCount is usually 1 in this case.
+    ///
+    /// The size of the provided workspace in Hybrid Find modes is often smaller than necessary for
+    /// Normal Find, because GWSS in these modes return size suitable only for the "best" solver
+    /// \ref ffind_gwss_why_not_0. If we check IsEnoughWorkspace() for all solvers, then many false
+    /// warnings may be produced. That is why we have to check IsEnoughWorkspace for the
+    /// maxSolutionCount "best" solvers only.
```
Comment (atamazov) [inline src/ocl/convolutionocl.cpp:789]: This explains the reason of the issue.

**Review by atamazov — COMMENTED** [review]
Notes for reviewers.

**Comment by atamazov** [comment]
## :cyclone: Routine performance/correctness testing

- Preconditions: Ubuntu 20.04.5 LTS, BUILD_DEV=On, Release build, Navi21/36, ROCm 6.0.2 docker
- Versions of MIOpen under test: `develop` 4347d7585 vs branch b15c949bf.
- Tested configs:
  - 93 known FP32 convolution configs
- Modes:
  - (1) Default Find mode (DYNAMIC_HYBRID) with default system find-db
  - (2) Immediate mode (with user find-db possibly updated at step (1))
- 🟢 No performance or correctness regressions.
- 🟢 Excessive warnings disappeared.

**Comment by atamazov** [comment]
## :cyclone: Feature correctness/performance

- Preconditions: Ubuntu 20.04.5 LTS, BUILD_DEV=On, Release build, Navi21/36, ROCm 6.0.2 docker
- Versions of MIOpen under test: `develop` 4347d7585 vs branch b15c949bf.
- Tested configs:
  - 93 known FP32 convolution configs
  - 93 known FP16 convolution configs
- Modes:
  - Default Find Mode, `MIOPEN_DRIVER_CONV_WORKSPACE_SIZE_ADJUST=2` (1/2 workspace), default find-db
  - Default Find Mode, `MIOPEN_DRIVER_CONV_WORKSPACE_SIZE_ADJUST=1` (Zero workspace), default find-db
- 🟢 No performance or correctness regressions.

**Review by junliume — APPROVED** [review]

**Pull request merged**